### PR TITLE
[WIP] Fix Ming text streaming TTFT

### PR DESCRIPTION
**Pull request by Hangzhi**
**For Reviewers** This PR is a first draft. Polish WIP. 

## Summary
Fixes Ming-Omni text streaming so OpenAI `stream=true` receives incremental `delta.content` chunks instead of one final full-text chunk.

Changes:
- route Ming thinker text deltas to `decode`
- add a stream-aware Ming decode scheduler that forwards text deltas to the coordinator
- make the final streamed decode result emit only any missing text suffix, avoiding duplication while preserving tail text
- preserve Ming/Bailing `head_dim` and `rotary_dim` config values when present
- align Ming attention tensor shape handling with the Bailing attention contract

## Measurement

Issue: https://github.com/sgl-project/sglang-omni/issues/600

See [Measurement](#measurement) below for the commands, methodology, and results.

OpenAI-visible TTFT benchmark, deterministic 8-token schedule:

| backend/source | TTFT mean | TTFT p95 | E2E mean | content chunks |
|---|---:|---:|---:|---:|
| sglang-omni main | 0.4519 s | 0.4521 s | 0.4520 s | 1 |
| sglang-omni current fix | 0.1006 s | 0.1007 s | 0.4540 s | 8 |

Source-backed synthetic/profile comparison against the local vLLM-Omni repo, using `prefill_s=0.0398`, `token_s=0.0102`:

| backend/source | TTFT mean | TTFT p95 | E2E mean | content chunks |
|---|---:|---:|---:|---:|
| sglang-omni current fix | 0.0515 s | 0.0516 s | 0.1310 s | 8 |
| sglang-omni main | 0.1296 s | 0.1298 s | 0.1297 s | 1 |

Interpretation: main waits until all generated text is available before emitting the first non-empty OpenAI stream chunk. The fix changes Ming text streaming to the same OpenAI-visible shape as vLLM-Omni source semantics: TTFT is prefill plus the first generated token, not full generation latency.

Note: `vllm` is not installed in this environment. The vLLM-Omni comparison above verifies the local vLLM-Omni source contracts (`RequestOutputKind.DELTA` for streaming and OpenAI `delta.content` SSE emission) and applies the reported vLLM-like timing profile. It is not a full vLLM-Omni server benchmark.

## Verification

```bash
timeout 90s /usr/bin/python3 -m pytest -q \
  tests/unit_test/ming_omni/test_pipeline.py \
  tests/unit_test/ming_omni/test_streaming_speech_config.py \
  tests/unit_test/ming_omni/test_streaming_e2e_glue.py \
  tests/unit_test/ming_omni/test_thinker.py

timeout 30s /usr/bin/python3 -m pytest -q \
  tests/unit_test/serve/test_openai_api.py::test_chat_stream_failure_closes_without_done_sentinel

timeout 30s /usr/bin/python3 -m pytest -q \
  tests/unit_test/qwen3_omni/test_streaming.py::test_client_completion_stream_does_not_duplicate_full_text \
  tests/unit_test/qwen3_omni/test_streaming.py::test_client_completion_stream_non_streaming_keeps_full_text

/usr/bin/python3 -m py_compile \
  sglang_omni/models/ming_omni/bootstrap.py \
  sglang_omni/models/ming_omni/config.py \
  sglang_omni/models/ming_omni/stages.py \
  sglang_omni/models/ming_omni/thinker.py \
  sglang_omni/models/ming_omni/configuration.py

git diff --check
```

Observed:
- `52 passed in 4.36s`
- `1 passed in 2.81s`
- `2 passed, 2 warnings in 10.37s`
- `py_compile` passed
- `git diff --check` passed

### CODE REVIEW

**Comment by edwingao28** [comment]
@Hangzhi please fix lint

File: sglang_omni/models/ming_omni/thinker.py [hunk 1]
```diff
@@ -102,9 +102,8 @@ def __init__(
             self.q_norm = RMSNorm(self.head_dim, eps=config.rms_norm_eps)
             self.k_norm = RMSNorm(self.head_dim, eps=config.rms_norm_eps)
 
-        # RoPE - using partial rotary factor
         self.rotary_emb = get_rope(
-            self.rotary_dim,
+            self.head_dim,
```
Comment (edwingao28) [inline sglang_omni/models/ming_omni/thinker.py:106]: **Attention/RoPE change is correct but unproven.** `get_rope`'s first positional arg is head_size, so `self.head_dim` is the right value here — `main` passed `self.rotary_dim`, which is wrong whenever `rotary_dim < head_dim` (partial rotary). Removing the manual reshape + split is also correct: `apply_qk_norm` reshapes by `head_dim` internally and `RotaryEmbedding` applies partial rotary via `rotary_dim`. So this is effectively a latent-bug fix — but it changes attention math for **every** Ming request (text + speech) and is guarded only by a source-string test (`test_ming_attention_keeps_qkv_flat_for_radix_attention` greps the source, it does not check numerics).

1. Please add a numerical forward-parity test (e.g. `cos ≥ 0.999` vs HF reference on a partial-rotary config), not a source-string assertion
2. Please either split the attention/RoPE/config-schema changes into a separate PR, or explain why they are required for the streaming fix in this PR.

File: sglang_omni/models/ming_omni/stages.py [hunk 2]
```diff
@@ -101,6 +110,255 @@ def _attach_decode_final_metadata(
...
+    ) -> None:
+        from sglang_omni.scheduling.messages import IncomingMessage, OutgoingMessage
+
+        self.inbox: _queue_mod.Queue[IncomingMessage] = _queue_mod.Queue()
+        self.outbox: _queue_mod.Queue[OutgoingMessage] = _queue_mod.Queue()
+        self._tokenizer = tokenizer
+        self._eos_token_id = eos_token_id
+        self.stage_name = stage_name
+        self._running = False
+        self._state: dict[str, _StreamingDecodeState] = {}
```
Comment (edwingao28) [inline sglang_omni/models/ming_omni/stages.py:215]: `MingStreamingDecodeScheduler` robustness — two gaps:

1. The per-message `try`/`except` in `start()` correctly isolates per-request failures (matches the Qwen3-Omni scheduler contract), but there is no test for it. Add one that makes `_on_stream_chunk`/`_finalize` raise and asserts the scheduler thread keeps processing later requests.
2. `self._state` has no bounded eviction (only `_done_seen` does). A request that streams chunks then is aborted before `new_request`/`stream_done` reaches decode leaks its `_state` entry forever → unbounded memory under failure traffic. Add a TTL/bounded eviction, or guarantee `abort()` on every termination path.

File: sglang_omni/models/ming_omni/stages.py [hunk 3]
```diff
@@ -101,6 +110,255 @@ def _attach_decode_final_metadata(
...
+                    return bytes(values).decode("utf-8", errors="ignore")
+                token_ids = [
+                    int(token_id) for token_id in data.detach().cpu().flatten()
+                ]
+                if token_ids:
+                    return self._tokenizer.decode(
+                        token_ids,
+                        skip_special_tokens=True,
+                    )
+        except Exception:
```
Comment (edwingao28) [inline sglang_omni/models/ming_omni/stages.py:356]: `except Exception: raise` is a no-op — it catches every exception only to immediately re-raise, identical to having no `try`/`except`. If the intent is to tolerate a missing torch, narrow to `except ImportError:` around just the `import torch` and `return str(data)`; otherwise drop the `try`/`except` entirely.

File: sglang_omni/models/ming_omni/bootstrap.py [hunk 4]
```diff
@@ -253,8 +263,10 @@ def make_thinker_stream_output_builder(
     tokenizer: Any,
     eos_token_id: int | None,
     target_stage: str = "segmenter",
+    target_stages: Sequence[str] | None = None,
```
Comment (edwingao28) [inline sglang_omni/models/ming_omni/bootstrap.py:266]: Dual public params `target_stage` (singular, default `"segmenter"`) + `target_stages` (plural) is a foot-gun: `target_stage` now only survives as a fallback when `target_stages is None`, and every internal call site passes `target_stages`. Drop `target_stage` (or wrap it into `target_stages` with a `DeprecationWarning`).

**Review by edwingao28 — CHANGES_REQUESTED** [review]

**Comment by edwingao28** [comment]
Please resolve file conflict and remove WIP in the title @Hangzhi, in general this fix PR looks good to me and I left some review comments.

**Comment by yuan-luo** [comment]
Please remove the vllm-omni benchmark in the description.

**Comment by edwingao28** [comment]
Thanks for the work here @Hangzhi . We decided to move forward with #656 as the merge path because it keeps the text streaming architecture a bit cleaner: thinker emits raw token ids and decode owns client-visible detokenization.

That said, #613 was very helpful in identifying the correct speech/streaming-speech topology. In particular, the idea that speech thinkers must stream completion to decode, and that decode must accept stream signals before payload, directly informed the final #656 changes, which I added you as co author.

I really appreciate the investigation and design work here.

**Comment by Hangzhi** [comment]
@edwingao28 sounds good. Let me close this PR in favor of https://github.com/sgl-project/sglang-omni/pull/656

Sorry that I did get the comments resolved earlier.
